engine/consensus/sealing/engine.go

```diff
@@ -359,49 +408,78 @@ func (e *Engine) Done() <-chan struct{} {
+	// randomness (to compute the chunk assignment). The source of randomness can be taken from _any_
+	// QC for the block. We know that we have such a QC, once a valid child block is incorporated.
+	// Vice-versa, once a block is incorporated, we know that _its parent_ has a valid child, i.e.
+	// the parent's source of randomness is now know.
 
-		// we are interested in blocks with height strictly larger than root block
-		if incorporatedBlock.Height <= e.rootHeader.Height {
-			return
-		}
+	incorporatedBlock, err := e.headers.ByBlockID(incorporatedBlockID)
+	if err != nil {
+		e.log.Fatal().Err(err).Msgf("could not retrieve header for block %v", incorporatedBlockID)
+	}
+
+	e.log.Info().Msgf("processing incorporated block %v at height %d", incorporatedBlockID, incorporatedBlock.Height)
+
+	// we are interested in blocks with height strictly larger than root block
+	if incorporatedBlock.Height <= e.rootHeader.Height {
+		return nil
+	}
+
+	index, err := e.index.ByBlockID(incorporatedBlock.ParentID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve payload index for block %v", incorporatedBlock.ParentID)
+	}
 
-		payload, err := e.payloads.ByBlockID(incorporatedBlock.ParentID)
+	for _, resultID := range index.ResultIDs {
+		result, err := e.results.ByID(resultID)
 		if err != nil {
-			e.log.Fatal().Err(err).Msgf("could not retrieve payload for block %v", incorporatedBlock.ParentID)
+			return fmt.Errorf("could not retrieve receipt incorporated in block %v: %w", incorporatedBlock.ParentID, err)
+		}
+		added := e.pendingIncorporatedResults.Push(result)
+		if !added {
+			// Not being able to queue an incorporated result is a fatal edge case. It might happen, if the
+			// queue capacity is depleted. However, we cannot dropped the incorporated result, because there
+			// is no way that an incorporated result can be re-added later once dropped.
+			return fmt.Errorf("failed to queue incorporated result")
 		}
+	}
+	e.inboundEventsNotifier.Notify()
+	return nil
+}
 
-		for _, result := range payload.Results {
-			added := e.pendingIncorporatedResults.Push(result)
-			if !added {
-				// Not being able to queue an incorporated result is a fatal edge case. It might happen, if the
-				// queue capacity is depleted. However, we cannot dropped the incorporated result, because there
-				// is no way that an incorporated result can be re-added later once dropped.
-				e.log.Fatal().Msg("failed to queue incorporated result")
+// processBlockIncorporatedEvents performs processing of block incorporated hot stuff events
+func (e *Engine) processBlockIncorporatedEvents() error {
+	for {
+		select {
+		case <-e.unit.Quit():
+			return nil
+		default:
+		}
+
+		msg, ok := e.pendingIncorporatedBlocks.Pop()
+		if ok {
+			err := e.processIncorporatedBlock(msg.(flow.Identifier))
+			if err != nil {
+				return fmt.Errorf("could not process incorporated block: %w", err)
 			}
+			continue
 		}
-		e.notifier.Notify()
-	})
+	}
```

Comment:
⚠️ I think you are missing a `return` statement here:
 ```suggestion
		return nil // no more events to process
	}
```
Otherwise, you'll never get out of this loop again.